src/garage/torch/policies/context_conditioned_policy.py

```diff
@@ -0,0 +1,283 @@
+        self.register_buffer('z_vars', torch.zeros(1, latent_dim))
+
+        self.reset_belief()
+
+    def reset_belief(self, num_tasks=1):
+        r"""Reset :math:`q(z \| c)` to the prior and sample a new z from the prior.
+
+        Args:
+            num_tasks (int): Number of tasks.
+
+        """
+        # reset distribution over z to the prior
+        mu = torch.zeros(num_tasks, self._latent_dim).to(tu.global_device())
+        if self._use_information_bottleneck:
+            var = torch.ones(num_tasks,
+                             self._latent_dim).to(tu.global_device())
+        else:
+            var = torch.zeros(num_tasks,
+                              self._latent_dim).to(tu.global_device())
+        self.z_means = mu
+        self.z_vars = var
+        # sample a new z from the prior
+        self.sample_from_belief()
+        # reset the context collected so far
+        self._context = None
+        # reset any hidden state in the encoder network (relevant for RNN)
+        self._context_encoder.reset(num_tasks)
+
+    # pylint: disable=attribute-defined-outside-init
+    def sample_from_belief(self):
+        """Sample z using distributions from current means and variances."""
+        if self._use_information_bottleneck:
+            posteriors = [
+                torch.distributions.Normal(m, torch.sqrt(s)) for m, s in zip(
+                    torch.unbind(self.z_means), torch.unbind(self.z_vars))
+            ]
+            z = [d.rsample() for d in posteriors]
+            self.z = torch.stack(z)
+        else:
+            self.z = self.z_means
+
+    def detach_z(self):
+        """Disable backprop through z."""
+        # pylint: disable=attribute-defined-outside-init
+        self.z = self.z.detach()
+
+    # pylint: disable=attribute-defined-outside-init
+    def update_context(self, timestep):
+        """Append single transition to the current context.
+
+        Args:
+            timestep (garage._dtypes.TimeStep): Timestep containing transition
+                information to be added to context.
+
+        """
+        o = torch.as_tensor(timestep.observation[None, None, ...],
+                            device=tu.global_device()).float()
+        a = torch.as_tensor(timestep.action[None, None, ...],
+                            device=tu.global_device()).float()
+        r = torch.as_tensor(np.array([timestep.reward])[None, None, ...],
+                            device=tu.global_device()).float()
+        no = torch.as_tensor(timestep.next_observation[None, None, ...],
+                             device=tu.global_device()).float()
+
+        if self._use_next_obs:
+            data = torch.cat([o, a, r, no], dim=2)
+        else:
+            data = torch.cat([o, a, r], dim=2)
+
+        if self._context is None:
+            self._context = data
+        else:
+            self._context = torch.cat([self._context, data], dim=1)
+
+    # pylint: disable=attribute-defined-outside-init
```

Comment:
please correct this by declaring this attribute in the constructor (either it has a sane default value, or your other methods will need to check that it has been initialized).